### PR TITLE
CloneFactory implementation for EIP 1167

### PR DESCRIPTION
**Pull request by pdyraga**
`CloneFactory` contract as used in https://github.com/keep-network/sortition-pools.

It is an implementation of EIP 1167 clone factory borrowed from https://github.com/optionality/clone-factory.

We use clone factory in `keep-network/sortition-pools` and we'll soon use it in `threshold-network/solidity-contracts` to create grant instances. Extracting it so that we do not have to copy-paste the code.

Simple unit tests for `CloneFactory` have been additionally implemented in this PR.

### CODE REVIEW

File: test/clone/CloneFactory.test.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        expect(
+          await cloneFactory.isClonePublic(
+            cloneable.address,
+            cloneable2.address
+          )
+        ).to.be.false
+      })
+    })
+
+    context("when contract is a clone of another contract", () => {
```
Comment (nkuba) [inline test/clone/CloneFactory.test.js:60]: Can we also add a test where we call `isClone` for clones of the same contract:

```js
    context("when both contracts are clones of the same contract", () => {
      it("should return true", async () => {
        const clone1 = await createClone()
        const clone2 = await createClone()
        expect(
          await cloneFactory.isClonePublic(clone1.address, clone2.address)
        ).to.be.true
      })
    })
```

**Review by nkuba — COMMENTED** [review]
All looks good, I just wanted to ask for one more test case.

File: test/clone/CloneFactory.test.js (hunk 1 above)
Comment (pdyraga) [reply test/clone/CloneFactory.test.js:60]: See https://github.com/thesis/solidity-contracts/pull/4/commits/73bc84f34565d85ccbaecc6ceb3f12ef6e84567b please.

File: contracts/clone/CloneFactory.sol [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+// Implementation of [EIP-1167] based on [clone-factory]
+// source code.
+//
+// EIP 1167: https://eips.ethereum.org/EIPS/eip-1167
+// clone-factory: https://github.com/optionality/clone-factory
+// Modified to use ^0.8.5; instead of ^0.4.23 solidity version
```
Comment (nkuba) [inline contracts/clone/CloneFactory.sol:31]: Can we improve the contract documentation? Keeping in mind https://github.com/thesis/solidity-contracts/pull/3#discussion_r677453642

Comment (pdyraga) [reply contracts/clone/CloneFactory.sol]: See d4812899e0cbfde4800fd65ab77613f8dd0308e4.

**Review by nkuba — APPROVED** [review]

**Pull request merged**